File: frontend/lib/models/report.dart
/// Mirrors the backend's ReportOut schema (app/schemas/reporting.py).
class Report {
  final int reportId;
  final String type;
  final String periodStart;
  final String periodEnd;
  final String status;
  final bool hasPdf;

  Report({
    required this.reportId,
    required this.type,
    required this.periodStart,
    required this.periodEnd,
    required this.status,
    required this.hasPdf,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      reportId: json['report_id'] as int,
      type: json['type'] as String,
      periodStart: json['period_start'] as String,
      periodEnd: json['period_end'] as String,
      status: json['status'] as String,
      hasPdf: json['pdf_path'] != null,
    );
  }
}

File: backend/app/models/report.py
"""
Module 11: Reporting & Analytics — Report [M11] (Ch4 §4.3.1). Backs
FR11.1-FR11.6.
"""
from datetime import date, datetime, timezone

from sqlalchemy import Date, DateTime, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class Report(Base):
    __tablename__ = "reports"

    report_id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    # weekly_summary | forecast_accuracy | waste_cost | budget_utilisation
    type: Mapped[str] = mapped_column(String(30), nullable=False)
    period_start: Mapped[date] = mapped_column(Date, nullable=False)
    period_end: Mapped[date] = mapped_column(Date, nullable=False)
    generated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), nullable=False
    )
    # Nullable until PDF export completes.
    pdf_path: Mapped[str | None] = mapped_column(String(255), nullable=True)
    # generated | failed
    status: Mapped[str] = mapped_column(String(20), default="generated", nullable=False)
